blog/tidy-finance-dummy-data/index.qmd

```diff
@@ -0,0 +1,422 @@
+
+```{r}
+number_of_stocks <- 100
+
+crsp_stocks <- tbl(tidy_finance, "crsp_monthly") |>
+  group_by(permno) |> 
+  filter(month == max(month, na.rm = TRUE)) |> 
+  ungroup() |> 
+  select(permno, gvkey, industry, exchange, exchcd, siccd) |> 
+  collect()
+
+industries <- crsp_stocks |> 
+  filter(industry != "Missing") |> 
+  count(industry) |> 
+  mutate(prob = n / sum(n))
+
+exchanges <- crsp_stocks |> 
+  filter(exchange != "Other") |> 
+  count(exchange) |> 
+  mutate(prob = n / sum(n))
+
+stock_identifiers <- 1:number_of_stocks |> 
+  map_df(
+    function(x) {
+      tibble(
+        permno = x,
+        gvkey = as.character(x + 10000),
+        exchange = sample(exchanges$exchange, 1, 
+                          prob = exchanges$prob),
+        industry = sample(industries$industry, 1, 
+                          prob = industries$prob)
+      ) |> 
+        mutate(
+          exchcd = case_when(
+            exchange == "NYSE" ~ sample(c(1, 31), n()),
+            exchange == "AMEX" ~ sample(c(2, 32), n()),
+            exchange == "NASDAQ" ~ sample(c(3, 33), n())
+          ),
+          siccd = case_when(
+            industry == "Agriculture" ~ sample(1:999, n()),
+            industry == "Mining" ~ sample(1000:1499, n()),
+            industry == "Construction" ~ sample(1500:1799, n()),
+            industry == "Manufacturing" ~ sample(1800:3999, n()),
+            industry == "Transportation" ~ sample(4000:4899, n()),
+            industry == "Utilities" ~ sample(4900:4999, n()),
+            industry == "Wholesale" ~ sample(5000:5199, n()),
+            industry == "Retail" ~ sample(5200:5999, n()),
+            industry == "Finance" ~ sample(6000:6799, n()),
+            industry == "Services" ~ sample(7000:8999, n()),
+            industry == "Public" ~ sample(9000:9999, n())
+          )
+        )
+    }
+  )
+```
+
+Next, we construct three panels of stock data with varying frequencies: yearly, monthly, and daily. We begin by creating the `stock_panel_yearly` panel. To achieve this, we combine the `stock_identifiers` table with a new table containing the variable `year` from `time_series_years`. The `crossing()` function ensures that we get all possible combinations of the two tables. After combining, we select only the `gvkey` and `year` columns for our final yearly panel.
```

Comment:
Small question: expand_grid() is not an alternative here? We use it in the book.

Reply:
crossing() is a wrapper around expand_grid() that de-duplicates and sorts its inputs. I hence prefer crossing() and I propose that we update the book accordingly.

Reply:
Actually, I changed my mind: let's keep expand_grid() because it is more verbose than crossing() and the difference is negligible.